### PR TITLE
Prompt the user to save or discard changes when leaving the new thread form

### LINKED ISSUE #519: When closing the editor, prompt whether the user should save changes as a new DiscussionDraft / save changes to the existing DiscussionDraft
(low priority - for after construct is merged)

We have two draft systems right now, DiscussionDrafts on the backend and localStorage on the frontend.

The localStorage system is important to have around (in case the editor or page crashes, or the user closes the window and we don't catch that) but it's a bit weird to have draft content hanging around in the editor all the time.

When the user closes the editor, we should prompt them to save the post's content to the backend as a DiscussionDraft, or discard the content. This should be the same interaction as the save draft button - the copy might need to subtly change depending on whether we are updating an existing draft or creating a new one - but as long as the user proceeds through the modal, we can clear localStorage, and let the user restore their previous work by loading it from the drafts panel next time.

### PR DESCRIPTION
**Pull request by gdjohnson**
Closes #519.

## Description
Previously, on closing the new thread form, all unsaved work would be saved to localStorage, and re-rendered on the next visit to the form. While this helped prevent unwanted loss of data in the event of e.g. a browser crash or accidental navigation, it was a little too heavy-handed: users would be confronted with old, unwanted text bodies or titles every time they went to create a post.

This branch narrows the scope of localStorage so that it has a lighter touch, saving form data past the modal-close only in the event of browser crashes or abruptly closed windows. Otherwise, when the new thread form is closed (as in modals) or navigated away from (as in the full page form), the user is prompted with a confirmation modal asking if they wish to save the thread as a draft (or, alternatively, to update their existing draft). Regardless of their response, the form data is cleared from localStorage (as the user has now had the opportunity to either save the form more permanently, as a discussion draft, or else discard it).

## How has this been tested?

This has been tested in both the modal form and full-page form, on both offchain and on-chain communities. I've QA'd user flows in which the user: 

1. opens the form, closes the form, and is not prompted
2. opens the form, adds title and/or text form data, then closes, and is prompted to 'Save draft?'
3. loads an existing draft, does not alter it, and is not shown a confirmation prompt
4. loads an existing draft, alters it, and is shown a confirmation prompt reading 'Update draft?'
5. on prompting, the user chooses to save/update the draft; it is successfully saved, and localStorage cleared
6. on prompting, the user does not save/update the draft; localStorage is cleared
7. the user abruptly exits the tab or window, and localStorage is not cleared
8. the user transitions from modal to full thread, and this "onremove" call does not trigger the modal

There are some lingering problems with the transition from modal to fullthread: we have to keep track of whether the draft has been altered, where the draft is coming from, the status of localStorage, etc—but this issue probably deserves its own issue.

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

## Does this PR affect any server routes?
- [ ] yes, and they are tested: [enter the % coverage here]
- [ ] yes, and they are not tested: [enter the % coverage here]
- [ ] yes, but I did not run tests
- [x] no

### CODE REVIEW

File: client/scripts/views/components/new_thread_form.ts [hunk 1]
```diff
@@ -159,16 +160,37 @@ export const NewThreadForm: m.Component<{
...
     }
-    if (vnode.state.newType === PostType.Discussion) {
+    if (vnode.state.postType === PostType.Discussion) {
       vnode.state.form.threadTitle = localStorage.getItem(`${app.activeId()}-new-discussion-storedTitle`);
     } else {
       vnode.state.form.url = localStorage.getItem(`${app.activeId()}-new-link-storedLink`);
       vnode.state.form.linkTitle = localStorage.getItem(`${app.activeId()}-new-link-storedTitle`);
     }
   },
+  onremove: async (vnode) => {
```
Comment (raykyri) [inline client/scripts/views/components/new_thread_form.ts:174]: Would this be smoother if we used onbeforeremove?

File: client/scripts/views/components/new_thread_form.ts [hunk 2]
```diff
@@ -159,16 +160,37 @@ export const NewThreadForm: m.Component<{
...
+        confirmed = await confirmationModalWithText(modalMsg)();
+        if (confirmed) {
+          await saveDraft(form, quillEditorState, null, fromDraft);
+          notifySuccess('Draft saved');
+        }
+      }
+      localStorage.removeItem(`${app.activeId()}-new-discussion-storedTitle`);
+      localStorage.removeItem(`${app.activeId()}-new-discussion-storedText`);
+      localStorage.removeItem(`${app.activeId()}-active-tag`);
+      localStorage.removeItem(`${app.activeId()}-post-type`);
```
Comment (raykyri) [inline client/scripts/views/components/new_thread_form.ts:191]: I feel like ideally the prompt should cancel the modal from closing if "Cancel" is pressed but that may be too difficul to do easily?

File: client/scripts/views/components/new_thread_form.ts (hunk 1 above)
Comment (gdjohnson) [reply client/scripts/views/components/new_thread_form.ts:178]: For whatever reason, onbeforeremove isn't hooking for me, have tried with several different configurations.

File: client/scripts/views/components/new_thread_form.ts (hunk 2 above)
Comment (gdjohnson) [reply client/scripts/views/components/new_thread_form.ts:195]: I'll switch up the copy for clarity

**Comment by raykyri** [comment]
"Save as draft" pops up after successfully creating a post. Otherwise QA seems good so far.

**Comment by gdjohnson** [comment]
> "Save as draft" pops up after successfully creating a post. Otherwise QA seems good so far.

Updated! I'm gonna PR the category_id branch really fast and then I can switch over to another round of QA on this.

Also, you may have noticed, but the changes to the modal --> full-page editor transition ended up being relatively minor—mainly ensuring that the full-page editor knew whether its contents had come from a draft or not. I ended up including those changes in this PR.

**Comment by raykyri** [comment]
Yep! Looks super smooth.

**Comment by gdjohnson** [comment]
Did another round, didn't catch anything. You?

**Comment by raykyri** [comment]
Seems good to me

**Pull request merged**